src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs

```diff
@@ -77,46 +77,67 @@ private static async Task IndexingAsync(ShellScope scope, IEnumerable<ContentCon
 
-                ContentItem published = null, latest = null;
-                bool publishedLoaded = false, latestLoaded = false;
+                ContentItem published = null;
 
+                if (context is PublishContentContext publishContext)
+                {
+                    if (publishContext.Cancel)
+                    {
+                        continue;
+                    }
+                    if (publishContext.PreviousItem != null)
+                    {
+                        published = publishContext.PreviousItem;
+                    }
+                }
                 foreach (var indexSettings in await luceneIndexSettingsService.GetSettingsAsync())
                 {
                     var cultureAspect = await contentManager.PopulateAspectAsync<CultureAspect>(context.ContentItem);
                     var culture = cultureAspect.HasCulture ? cultureAspect.Culture.Name : null;
                     var ignoreIndexedCulture = indexSettings.Culture != "any" && culture != indexSettings.Culture;
-
+ 
                     if (indexSettings.IndexedContentTypes.Contains(context.ContentItem.ContentType) && !ignoreIndexedCulture)
                     {
-                        if (!indexSettings.IndexLatest && !publishedLoaded)
-                        {
-                            publishedLoaded = true;
-                            published = await contentManager.GetAsync(context.ContentItem.ContentItemId, VersionOptions.Published);
-                        }
-
-                        if (indexSettings.IndexLatest && !latestLoaded)
-                        {
-                            latestLoaded = true;
-                            latest = await contentManager.GetAsync(context.ContentItem.ContentItemId, VersionOptions.Latest);
-                        }
-
-                        var contentItem = !indexSettings.IndexLatest ? published : latest;
-
-                        if (contentItem == null)
+                        if (context is RemoveContentContext)
                         {
                             await luceneIndexManager.DeleteDocumentsAsync(indexSettings.IndexName, new string[] { context.ContentItem.ContentItemId });
+                            continue;
                         }
-                        else
+                        var stroeLuceneDocument = async (ContentItem contentItem) =>
                         {
                             var buildIndexContext = new BuildIndexContext(new DocumentIndex(contentItem.ContentItemId, contentItem.ContentItemVersionId), contentItem, new string[] { contentItem.ContentType }, new LuceneContentIndexSettings());
                             await contentItemIndexHandlers.InvokeAsync(x => x.BuildIndexAsync(buildIndexContext), logger);
-
-                            await luceneIndexManager.DeleteDocumentsAsync(indexSettings.IndexName, new string[] { contentItem.ContentItemId });
+                            if (!indexSettings.IndexLatest)
```

Comment:
I think here it should be the opposite. Here IndexLatest means that we want to index draft contentItems. We should need to delete by contentItemVersionId only when IndexLatest is set to true.